### ISSUE TITLE
Automatic record module application

### ISSUE THREAD

**Issue by jliptrap** [agda/agda#5613]
As a toy example, please consider the following two ways of defining a magma homomorphism:

```agda
open import Agda.Primitive
open import Agda.Builtin.Equality

record Magma : Set₁ where
  field Obj : Set
  field _*_ : Obj -> Obj -> Obj

record UnqualifiedStyleHomom (X Y : Magma) : Set where
  open Magma

  field run : X .Obj -> Y .Obj
  field compat : (x x' : Obj X) -> Y ._*_ (run x) (run x') ≡ run (X ._*_ x x')

record QualifiedStyleHomom (X Y : Magma) : Set where
  private module X = Magma X
  private module Y = Magma Y

  field run : X.Obj -> Y.Obj
  field compat : (x x' : X.Obj) -> run x Y.* run x' ≡ run (x X.* x')
```

I find the qualified style more readable except for the `private module` boilerplate lines. Could the compiler be so kind as to effectively autogenerate those lines?  E.g., when the user writes `X.*`, if `X` has record type but no module `X` is in scope, the compiler could apply the corresponding record module to `X` (as newcomers to Agda presumably expect).

IIUC this feature would be backwards compatible, and may address #2587 and to some extent #632.

**Comment by andreasabel** [comment]
> when the user writes `X.*`, if `X` has record type but no module `X` is in scope, the compiler could apply the corresponding record module to `X`

This seems to be a very high hanging fruit given the current architecture of Agda that separates a scope-checking pass from the later type-checking pass.  The scope checker does not know that `X` and `Y` are of record type, as it has no concept of "type". 

Maybe what could be done is to allow `X.f` as syntax for post-fix projection application `X .f`.  This would then enable the infix use as in `X.*`.  
Currently though, `X.f` can only mean the name `f` from module `X`, and the scope checker can resolve it.  If `X` is not a module, then the scope checker cannot disambiguate `f`.  Also, giving `X.f` a new interpretation would not be backwards compatible, because it would introduce an ambiguity:
```agda
record R : Set1 where
  field
    f : Set

module M where
  postulate
    f : Set
open R

postulate
  M : R

A = M .f  -- This is field f of record M.
B = M.f   -- This is declaration f of module M.
```
As we can see here, `M .f` and `M.f` can be different things.

**Comment by jliptrap** [comment]
Currently I avoid post-fix projections due to #2587; I prefer a uniform convention for record fields and record methods.  If post-fix projections supported record methods and infix notation (but not arbitrary functions), I think that would be good enough:

```agda
record PostfixProjectionStyleHomom (X Y : Magma) : Set where
  open Magma

  field run : X .Obj -> Y .Obj
  field compat : (x x' : X .Obj) -> run x (Y .*) run x' ≡ run (x (X .*) x')
```

I'm not familiar with Agda internals but it makes sense to me that my original proposal might involve too much compiler complexity to be worth it.  In the limit, of course, I would prefer that the compiler read my mind, figure out what theorem I'd like to prove, and prove it, so that I wouldn't have to type a single character.  :)

**Comment by andreasabel** [comment]
Using postfix-projections as infix operators as in `run x (Y .*) run x'` could maybe made to work in the operator parser.  However, this would not be backwards-compatible.  Note that the following is legal Agda:
```agda
record Magma : Set₁ where
  field
    *   : Set
    _*_ : * → * → *
open Magma

postulate
  M   : Magma
  F   : M .* → Set → M .* → Set
  x y : M .*
  foo : F x (M .*) y
```
This is because for Agda `*` and `_*_` are different identifiers.

**Comment by jliptrap** [comment]
Alternatively, if post-fix projections will not support record methods, I think it would be very useful to have sugar for the heavy repetition:
```agda
  private module X = Magma X
  private module Y = Magma Y
```
Maybe something along the lines of `open Magma at (X ; Y)`?

**Comment by andreasabel** [comment]
These declarations do not open any module, they abbreviate module applications.
You can shorten this to
```agda
  private
    module X = Magma X
    module Y = Magma Y
```
I agree that this is still noisy, but I do not see any principled way of shortening this.

You could look for solutions using macro preprocessors if you find this too noisy.

**Comment by jliptrap** [comment]
Having users duct tape macro preprocessors onto the Agda compiler would be lamentable from a systems perspective, e.g., collaboration, editor tooling, and robustness.

Is there a non-noisy way to use infix record methods (e.g., subtraction for an abelian group)?

**Comment by andreasabel** [comment]
> Is there a non-noisy way to use infix record methods (e.g., subtraction for an abelian group)?

You can use `open` with `renaming`.  E.g., something close to your `QualifiedStyle`:
```agda
record PostfixProjectionStyleHomom (X Y : Magma) : Set where
  open Magma X renaming (Obj to X::Obj; _*_ to _X::*_)
  open Magma Y renaming (Obj to Y::Obj; _*_ to _Y::*_)

  field run : X::Obj -> Y::Obj
  field compat : (x x' : X::Obj) -> run x Y::* run x' ≡ run (x X::* x')
```

Personally, I prefer a style where records are not opened, only modules:
```agda
record Magma : Set₁ where
  field Obj : Set
  field _⊢_*_ : Obj → Obj → Obj

record QualifiedStyleHomom (X Y : Magma) : Set where
  open Magma

  field run : X .Obj → Y .Obj
  field compat : (x x' : Obj X) → Y ⊢ run x * run x' ≡ run (X ⊢ x * x')
```
Here, you always state in which structure you operate, via the turn-style prefix.

**Comment by jliptrap** [comment]
Renaming style is of course the noisiest of all, but turnstile style is a nice hack, thanks!  IIUC it's still noisy because it tends to require doppelgängers within the record module:

```agda
record Monoid : Set₁ where
  field Obj : Set

  field id : Obj
  field _*_ : Obj → Obj → Obj
  _⊢_*_ = _*_

  field assoc : (a b c : Obj) -> (a * b) * c ≡ a * (b * c)
```

**Comment by andreasabel** [comment]
This doppelgänger trick is new to me, shall remember it!

I just realized that the turnstile-trick also goes well with definition by copatterns:
```agda
record Magma : Set₁ where
  field Obj : Set
  field _⊢_*_ : Obj → Obj → Obj
open Magma

natMagma : Magma
natMagma .Obj    = Nat
natMagma ⊢ x * y = x + y
```

**Comment by andreasabel** [comment]
Closing this issue...  
For a language change, we need a technical proposal.  There is none in sight here, seems very difficult.

**Comment by JasonGross** [comment]
The turnstyle proposal does not work well with nesting operations, `X ⊢ a * (X ⊢ b * (X ⊢ d * e))` is kind-of noisy, IMHO.
